### PR TITLE
Replace CoRN's license by Coq's license

### PR DESCRIPTION
**Pull request by VincentSe**
For better interactions with other Coq libraries, use the same license as Coq, which is LGPL 2.1. This pull request simply replaces CoRN's LICENSE file by Coq's LICENSE file.

### CODE REVIEW

**Comment by Zimmi48** [comment]
For the reasons I gave in the related PR, I think that if an effort is made to relicense CoRN, the goal should be a more permissive license than LGPL.

**Comment by VincentSe** [comment]
@Zimmi48 Because CoRN `Require Import` Coq's standard library, CoRN is forced to be LGPL. Or am I missing something ?

**Comment by Zimmi48** [comment]
Fortunately, it doesn't work like this. But you are perfectly illustrating my point on the LGPL being a complex license to understand. (I can come back to this and write a—long—explanation if you wish.)

**Comment by VincentSe** [comment]
@Zimmi48 Who is interpreting the license now ? You say that LGPL is complicated so we should avoid it, but eventually CoRN imports files from the standard library. If there is a problem with LGPL, by using the MIT license you would wrongly let users assume more permissions than they actually have. What I propose is much simpler : one license for both Coq and CoRN.

**Comment by Zimmi48** [comment]
>Who is interpreting the license now ?

I am. But I'm relatively confident in my interpretation of standard open source licenses based on [my experience reading about them and discussing them](https://opensource.stackexchange.com/users/5858/zimm-i48).

>by using the MIT license you would wrongly let CoRN users assume more permissions than they actually have

Let us be specific. No open source license (be it MIT, LGPL or GPL) places any restriction on how you can *use* a piece of software. Software licenses that do place restrictions on usage (like [CompCert's Non-Commercial License](https://github.com/AbsInt/CompCert/blob/master/LICENSE)) are non-free and non-open source. Open source licenses restrict how you may *redistribute* the software or a derivative.

For instance, with a copyleft license (like the LGPL) you may not distribute a binary version without also making the sources available, and if you create a derived version, you may be bound by the same restrictions.

If I distribute Coq + CoRN, of course, I am bound by the terms of Coq's license. However, if I distribute only CoRN, I am only bound by the terms of CoRN's license. So it makes a difference.

>What I propose is much simpler : one license for both Coq and CoRN.

I agree that what you propose is very simple. However, my point (which was previously raised by @robbertkrebbers) is that if you go through the burden of relicensing CoRN, you may as well go one step further and choose a more permissive license. If some day, Coq's standard library itself is relicensed under a more permissive license (which would be difficult but not impossible), it would be sad to have to conduct again the same relicensing process for CoRN. Choosing a permissive license that allows relicensing (like MIT or BSD) now guarantees against such issue.

**Comment by VincentSe** [comment]
> I agree that what you propose is very simple.

@Zimmi48 Great, let's keep it simple. This is complicated enough already.

> If some day, Coq's standard library itself is relicensed under a more permissive license (which would be difficult but not impossible)

This is not happening, you know it as much as I do.

**Comment by Zimmi48** [comment]
>This is not happening, you know it as much as I do.

Well, I don't know it, so I wonder how you can be so sure.

>Great, let's keep it simple. This is complicated enough already.

Well in any case, I'd oppose to relicensing to LGPL-2.0-only. At the very least, make this LGPL-2.0-or-later.

**Comment by VincentSe** [comment]
@Zimmi48 The option to use a later version of the license is already given in each file

```
 * This work is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
```

By the way, that means we need to modify all the files, replacing "GNU General Public License" by "GNU Lesser General Public License".

I do not know how to modify `meta.yml` what should I put in it ? Then how do we regenerate the `README` and opam file ?

**Comment by Zimmi48** [comment]
> By the way, that means we need to modify all the files, replacing "GNU General Public License" by "GNU Lesser General Public License".

That's correct. And "version 2" should become "version 2.1".

> I do not know how to modify `meta.yml` what should I put in it ? Then how do we regenerate the `README` and opam file ?

You only need to change two lines:

https://github.com/coq-community/corn/blob/9db8c9e8919beb718475c5005e06730d9a24fbb7/meta.yml#L75-L77

The identifier should become `LGPL-2.1-or-later`.

Then files can be regenerated using:

```
cd <your_coq_project> && cd ..
git clone https://github.com/coq-community/templates.git
cd -
../templates/generate.sh
```

(cf. https://github.com/coq-community/templates/blob/master/README.md)

But it would be best to merge #97 before doing this. Can you answer my question over there?

**Comment by VincentSe** [comment]
@Zimmi48 Is it good now with meta.yml modified and the templates regenerated ?

**Comment by Zimmi48** [comment]
Note that you probably had a slightly outdated version of the templates, leading to the (spurious) change to `.travis.yml`, which is responsible for CI failing (again).

**Comment by VincentSe** [comment]
@Zimmi48 I regenerated the templates and `.travis.yml` is back. It still does not build though. `opam install coq-corn.dev -v -y -j 2' failed`.

File: coq-corn.opam [hunk 1]
```diff
@@ -40,10 +40,7 @@ CoRN includes the following parts:
   integrals, graphs of functions, differential equations.
 """
 
-build: [
-  ["./configure.sh"]
-  [make "-j%{jobs}%"]
-]
+build: [make "-j%{jobs}%" ]
```
Comment (Zimmi48) [inline coq-corn.opam:43]: This is because of this change. Unfortunately, the coq-community templates do not support yet the addition of this `./configure.sh` step, so you have to restore it and only maintain the line change regarding the license identifier.

Comment (VincentSe) [reply coq-corn.opam]: Done

**Comment by Zimmi48** [comment]
This time it is the Nix builds which are failing because of the changes to `default.nix`. You should backtrack on those changes too. Note that the opam file should still have its license identifier modified.

**Comment by VincentSe** [comment]
@Zimmi48 Now branch master builds, but not the previous branches
`make: *** No rule to make target 'install'.  Stop.`

**Comment by Zimmi48** [comment]
Cf. my comment just above.

**Comment by herbelin** [comment]
> I agree that what you propose is very simple. However, my point (which was previously raised by @robbertkrebbers) is that if you go through the burden of relicensing CoRN, you may as well go one step further and choose a more permissive license.

Would it make sense to have multi-licensing? This seems to be considered in some [situations](https://en.wikipedia.org/wiki/Multi-licensing#License_compatibility).

More generally, why one licence for the whole repository, rather than one (or more) licence per file (if not per Coq declaration). If the problem is a problem of compatibility, we could treat it with the same kind of granularity as we treat e.g. option `--impredicative-set`, couldn't we???

**Comment by spitters** [comment]
Multi licensing could be a possibility, but there doesn't seem to be a reason to go for LGPL, over MIT/BSD.
So, I agree with @Zimmi48 and @robbertkrebbers that the latter option would be more future proof.

**Comment by herbelin** [comment]
Thinking at this licensing issue, there is something I still don't understand.

Here, we are only talking about source code, while combining license is about building binaries, right? For instance, according to GPL's FAQ [What is the difference between “mere aggregation” and “combining two modules into one program”?](https://www.gnu.org/licenses/old-licenses/gpl-2.0-faq.en.html#MereAggregation), two binaries which interact through a communication channel rather than being linked together can be combined whatever the license of the other is if one is GLP.

So, having Coq source code in different licenses such as GPL, LGPL or MIT seems to be no problem. The burden is put on who wants to make a binary executable out of the code having different licenses: they will have to make the components whose licenses are incompatible to communicate over a channel rather than by linking.

Is this interpretation correct? And if yes, is this constraint put on who wants to create a binary out of Coq code so critical?

One may also wonder whether distributing source code should satisfy that the licenses of every code present in the distributed archive have to be compatible. Still according to the GPL FAQ above, it is ok to put side by side a GPL program with a distinct program having any other license on the same CD-ROM. So, it is a fortiori ok to do so for the sources of the programs (and we actually already do it for the platform).

TL;DR: From the Coq point of view, I don't see problems with incompatible licenses other than forcing people interested in producing binaries to have the incompatible binaries in separate processes, and, for compatible licenses, to adopt the license resulting of this compatible match.

The remaining questions are then "political". Do we want to force people possibly producing binaries (such as a certified compiler with a floating point component) to have these libraries communicating over a channel rather than being linked? In practice, it is not even sure that it matters, since the executable part of CoRN, which is the part liable to be linked in a binary, is already MIT. Then, the actual licence incompatibilities I know regarding real numbers are with Flocq and Coquelicot, but, since there are less developers involved here, it may be easier to change the license of these packages, say, from LGPL 3 to LGLP 2 or any later version so that any combination of code related to real or floating-point numbers become acceptable as part of a same binary.

**Comment by Zimmi48** [comment]
If someone wants to build a certified piece of software they may be interesting to depend on more than one Coq library. If they do so, then they create a derivative work of the various libraries they picked. If these libraries' licenses are not compatible, they won't be able to do so. It's as simple as this. Discussing workarounds like having two independent programs communicate on a textual channel doesn't make sense in a software verification context IMHO.

Furthermore, no, combining is not just about binaries. If it were, then programs in interpreted languages wouldn't be submitted to the same kind of constraints than the rest. It would also mean that people could write proprietary code that depends on a GPL'd library and distribute the two side by side (as source code) and put whatever restrictions they want on the code they wrote.

**Comment by spitters** [comment]
Corn has had MIT code in the fast/ directory since 2007.

**Comment by herbelin** [comment]
@Zimmi48, I'm concerned by the current CoRN situation but I don't have a strong experience with licenses, my apologies. My feeling, especially after Abhishek asked his license question recently on Coq-Club, was that there was no clear consensus on what a software license should exactly imply in the context of formal proofs. Is there some jurisprudence to rely on?

So, regarding the MIT-licensed part of CoRN, should we consider that it is a non-sense to have part of the MIT-licensed directory of CoRN to depend on the part of CoRN marked as GPL without being itself GPL?

**Comment by Zimmi48** [comment]
> should we consider that it is a non-sense to have part of the MIT-licensed directory of CoRN to depend on the part of CoRN marked as GPL without being itself GPL?

No, I believe that it is OK to put some part of the code under a more permissive (and *GPL-compatible*) license, as long as it is clear to everyone that the "larger work" resulting from the combination of the two differently-licensed parts is GPL licensed.

**Comment by herbelin** [comment]
Then, what is the problem in distributing the GPL 2 part of CoRN, the MIT part of CoRN, and the LGPL 2.1 stdlib as part of a same repository? The constraint is only on who wants to "combine" the different parts?

As for the case of the platform, is it ok to distribute an archive which contains both GPL 2 and LGLP 3, even though they are incompatible?

**Comment by Zimmi48** [comment]
> Then, what is the problem in distributing the GPL 2 part of CoRN, the MIT part of CoRN, and the LGPL 2.1 stdlib as part of a same repository? The constraint is only on who wants to "combine" the different parts?

There is no problem at all. It's just that if one wanted at some point to move some CoRN code to the Coq repository, it wouldn't be possible if that code is under GPL. Furthermore, managing several licenses in a single repository creates more maintenance burden (you cannot easily move code between the various parts) and more confusion for the users.

>As for the case of the platform, is it ok to distribute an archive which contains both GPL 2 and LGPL 3, even though they are incompatible?

Yes, this is not a problem. In this case, it would really be considered as side-by-side distribution, pretty similarly to your earlier CD-ROM example. Of course, problems arise if users want to use several libraries that are incompatible license-wise in a single project, so this should still be a concern for the platform maintainers. In practice, I don't know of any such examples since CoRN is the only Coq library I know which is under GPL and it is not under GPL-2-only, it is under GPL-2-or-later as testified by the headers of those files under GPL.

**Comment by herbelin** [comment]
> it is under GPL-2-or-later as testified by the headers of those files under GPL.

Ah right. I looked at the main LICENCE file which says GPL 2, and I did not see the "or later" in each file.

> Yes, this is not a problem. In this case, it would really be considered as side-by-side distribution, pretty similarly to your earlier CD-ROM example. Of course, problems arise if users want to use several libraries that are incompatible license-wise in a single project

This is exactly the kind of thing which confuses me. Since we are only talking about files with full source information, where is the limit between "side-by-side" and a "single project"? This seems to be rather a matter of convention. We could organize things as CoRN-A and CoRN-B, or Platform-CoRN and Platform-Flocq, or Platform-CoRN-A, Platform-Corn-B and Platform-Flocq. It seems in last analysis to be only about how "people" perceive things as going together or not. Am I missing something?

Similarly, if it is ok to have both MIT-licensed and GNU-licensed code in the CoRN library, why couldn't it be possible to mix licenses also in the Coq repository?

**Comment by Zimmi48** [comment]
>It seems in last analysis to be only about how "people" perceive things as going together or not.

Ultimately, yes, interpreting law, contracts and licensing is about how people perceive things. However, there is a key difference here between distributing a set of packages in a single bundled and creating a project that needs to import two libraries to be evaluated. Only the later kind is an example of a "larger work".

>Similarly, if it is ok to have both MIT-licensed and GPL-licensed code in the CoRN library, why couldn't it be possible to mix licenses also in the Coq repository?

The only limitation would be the one that we, as Coq maintainers, choose to impose. I would personally be OK with accepting that clearly defined and limited components are put under a more permissive license (in fact that was the case for `coq_dune.ml` in prevision of its integration into Dune itself, and the reference manual is still under a different license than Coq). However, I would be opposed to introducing code under a less permissive license into the standard library.

**Comment by herbelin** [comment]
> However, I would be opposed to introducing code under a less permissive license into the standard library.

I can understand this, but how much does this matter in practice if it can just be instead two repositories or a repository of repositories?

By the way, is the licence of camlp5 LGPL-compatible for instance?

For instance, to be extreme, why don't we attach license to theorems (as proposed above), reflecting what the contributor of the theorem intends. As far as I understood, this is actually what basically happened to CoRN. If I understood correctly, mostly nobody cared about a license, except two persons. One proposed to use GPL and the others agreed because they did not care. But another (more recent) contributor (then followed by a group of developers) used MIT for his (their) own work.

Told otherwise, as soon as what matters is source code, which is the case for us, if we would not want to do politics regarding how this source code might eventually be used, we may also welcome contributors as they are and accept any license for any (distinct) contribution. (I'm not saying that we don't want to do politics though!)

**Comment by spitters** [comment]
@herbelin your description of the history is correct. In fact, I'm aware of only one strong GPL proponent, who has since long left academia.

**Comment by Zimmi48** [comment]
> By the way, is the licence of camlp5 LGPL-compatible for instance?

FYI, the license of camlp5 is what is known as 3-clause BSD (BSD3 for short and BSD-3-Clause in the SPDX classification) and yes, [it is compatible with the (L)GPL licenses](https://www.gnu.org/licenses/license-list.html#ModifiedBSD).

>For instance, to be extreme, why don't we attach license to theorems (as proposed above), reflecting what the contributor of the theorem intends.

Most people have no idea what licenses entail, and as I explained above, the more licenses you get, the more complicated you make the life of both users and maintainers. Of course, in an extreme world like the one you propose, one would develop additional tools to make license management bearable and would actually limit the choice of licenses to have a gradation but no equivalent and incompatible licenses. Anyway, the most reasonable approach which a lot of people support today is that it's good to consolidate an open source community around a single open source license (see the first proposition of https://choosealicense.com/). Unfortunately, the historically preferred license of the OCaml (and thus Coq) communities have been LGPL-2.1(-only) which as I've explained [somewhere else](https://github.com/coq-community/templates/issues/34#issuecomment-634565291) is a technically complex license that is not the best suited for Coq (nor OCaml) libraries. Given that the community is far from united around this license anyway, I think it make sense to start actively promoting other licenses such as the permissive (and very popular) MIT license or the weak copyleft (but simpler to understand than LGPL) MPL-2.0.

@spitters 
>I'm aware of only one strong GPL proponent, who has since long left academia.

But will this author agree to relicense their work under a more permissive license? Otherwise, this whole relicensing enterprise is bound to be never complete.

**Comment by spitters** [comment]
He seems to have disappeared from the internet. There are no papers
describing substantial contributions to the corn library.

**Comment by Zimmi48** [comment]
Does this mean that you would remove his code from CoRN?

**Comment by lmamane** [comment]
> @Zimmi48 Because CoRN `Require Import` Coq's standard library, CoRN is forced to be LGPL. Or am I missing something ?

You are inverting the situation. Actually, it is the GPL of CoRN that "forces" the combination of Coq and CoRN to be GPL, under two different possible mechanisms (any person distributing such a combination can choose which mechanism to use).

Under section 6 of the LGPLv2.1, one is allowed to link an LGPL-covered work (such as Coq and/or its "theories" standard library) with any other work (such as CoRN), and redistribute the combination, with some quite weak conditions on the licence of the "other work". The GPL fulfils these weak conditions. In effect, the license of the result of the combination of Coq and CoRN is the GPL. ("Obviously" if one takes out Coq without CoRN from the combined work distributed under section 6, then that copy of Coq can be used under the LGPL.) This is the same mechanism that allows a closed-source program, distributed as binary executable code only, to link dynamically against an LGPL-covered library. Or a free-as-in-free-speech / open source software under a GPL-incompatible license to do the same.

Another path for this specific case is that LGPL section 3 allows any person having a copy of Coq to redistribute _that copy_ under the GPL. Then you have a GPL Coq and a GPL CoRN, and the combination is covered by the GPL. That path is specific to GPL, it doesn't work for any other license.

**Comment by lmamane** [comment]
> I agree that what you propose is very simple. However, my point (which was previously raised by @robbertkrebbers) is that if you go through the burden of relicensing CoRN, you may as well go one step further and choose a more permissive license. If some day, Coq's standard library itself is relicensed under a more permissive license (which would be difficult but not impossible), it would be sad to have to conduct again the same relicensing process for CoRN. Choosing a permissive license that allows relicensing (like MIT or BSD) now guarantees against such issue.

To guard against such issue, no need to relicense to a non-copyleft license. You only need to change the license of CoRN to "Lesser GPL v2.1 or later, or, at your option, the license of the Coq standard library (theories directory as of time of writing)". Provided all copyright holders agree, that is not more work than relicensing to "Lesser GPL v2.1 or later". In other words, "Lesser GPL v2.1 or later, with the added permission that you can also choose the same licence as the Coq Standard Library (theories directory)". In effect the copyright holders say they trust the Coq copyright holders to make good choices in the future, and delegate to them collectively together (not individually) the power to relicense CoRN.

A more complicated setup which works in other scenarios is to give that power to an identified institution or group of persons... The Free Software Foundation, the Radboud University, the group "Bas Spitters and Robbert Krebbers and Herman Geuvers", the CNRS, the NWO or whatever the copyright holders collectively choose. I don't think that is a good setup for CoRN, but if that's what other people want, I'm willing to consent to a reasonable identified institution or group of people.

**Comment by lmamane** [comment]
> here doesn't seem to be a reason to go for LGPL, over MIT/BSD.

Well, this harks back to the question "why is CoRN under the GPL in the first place?" and the related "why is Coq under the Lesser GPL, rather than MIT/BSD/..."? For me, it is linked to a certain view of scientific progress and the role you have (and I had at the time) as academics. That things are done incrementally, by standing on the shoulder of giants, and we add the common pot of humanity's advancement and knowledge. And whether, or not, people should be allowed to make use of that common pot without contributing back their work to that same common pot. That used to be "obvious"; Newton and Leibnitz never had to worry whether someone making improvements on the calculus the developed would use the improved calculus for personal gain, but forbid others from further improving the improved calculus. Bohr and Heisenberg didn't have to worry that people making refinements to their Copenhagen interpretation of quantum mechanics would forbid other people to make further refinements. The idea would have seemed outlandish, bizarre and just not possible.

We now live in a world where that is not obvious anymore. And thus, as far as we intend our work to be in an ecosystem of "common pot", we have to use mechanisms like copyleft to express that. There is a strong copyleft (the GPL) and a weak copyleft (the Lesser GPL). The argument, which is being made, that CoRN would be better served by being under a weak copyleft doesn't mean we have to abandon all expression of the common pot ecosystem, and go all the way to no copyleft at all.

Secrecy of knowledge is so abhorrent to humanity and progress of humanity, that about a century (and a half?) ago, when secrecy of new ideas was a concrete problem, leading to loss of progress when people died, humanity devised a system, and agreed to make sacrifices, to entice inventors of new ideas to make them public, through the patent system: give the inventor a _temporary_ monopoly on usage of his idea, in exchange for making the idea public. That was the basic deal. Under the same basic position (abhorrence of secret of ideas leading to progress), we use copyleft to discourage secrecy.

**Comment by herbelin** [comment]
> FYI, the license of camlp5 is what is known as 3-clause BSD (BSD3 for short and BSD-3-Clause in the SPDX classification) and yes, [it is compatible with the (L)GPL licenses](https://www.gnu.org/licenses/license-list.html#ModifiedBSD).

OK, thanks.

Following  @lmamane's comments, I would like to add two comments.

To mitigate the situation, it is probably worth to replace the discussion in an historical context which changed in the last 20 years. At some time, part of the academic world was pretty irritated by the closed attitude of some big software companies (as witnessed by critical books and posts at this time). That justified, in reaction to these attitudes, to strongly defend strict copyleft policies. Some of these companies learnt from it and started to support open source software (e.g. acquiring GitHub), reducing this feeling of a conflict between proprietary and open source software. On the other side, the emergence of startups in domains close to us, and sometimes from colleagues, contributed to the idea that it is legitimate for such startups to protect their source, at least temporary, so that they can simply live from their work. Another aspect whose impact on the current ways of thinking I'm not able to quantify is the role of "permissive" approaches to licensing such as the one defended by the widely spreaded interpreted language Python.

My second comment is a question: do we have a clear idea of the impact of a license in our field?

At the technical level, I see at least two situations where an entity might want to keep a Coq development secret.

  1. A certification proof can be a commercial argument and some actors may want to announce that they prove some correctness properties without releasing the proof (this happened e.g. for the certification of JavaCard properties). In this case, a certification agency serves as third-party trustee. But in any cases, the proofs are not distributed, so the license of the public code possibly used in the certification has a priori no impact, right?

  2. A Coq development is used by extraction to produce a software (e.g. a certified compiler, or certified server). If the dependencies of this software are in a permissive license, the software can be distributed without releasing the source without thinking further. If some (unmodified) dependencies are in a strong copyleft license (GPL), some constraints will be put on the distributed executable if the author would not like to distribute the sources. Typically, the users will be asked to download and extract independently the strong copyleft Coq dependencies of the executable to make themselves a fully operational software. If this way of doing is compatible with the words of strong copyleft (and the users ok to do the combining themselves), this is a burden, but not a so strong one.

In the latter, I neglect the moral aspect. If the intent of the author of the strong copyleft license is to force works depending on it to be released with sources, it might be unwelcome to bypass this intent and be (morally) dommageable to release without sources. On the other side, the strong copyleft might be considered too constraining for who wants to release executables without releasing the sources and move them to rely on other, more permissive, dependencies. Yet on another side, good ideas kept secret do not remain secret for long. For instance, proprietary spreadsheets or word processors did not so much prevent the development of open-source similar softwares. Proprietary drivers did not significantly prevent developing open-source versions of these drivers. This is a question of time, and a natural swing of the pendulum, which we can see in other situations (e.g. the current reactions to an excess of tracking of internet users).

At the end, my feeling is that what really matters is not the technical impacts but the moral impact (and a regular attention that no actors abuse their power).

Back to the PR, I see that several people would feel more comfortable with sending the message of a more permissive license, either MIT-like, or LGPL 2.1, or "Lesser GPL v2.1 or later, or, at your option, the license of the Coq standard library (theories directory as of time of writing)". Do we then try to go for the latter, which after all seems reasonable? Then, an appropriate mail to CoRN developers could be prepared.

**Comment by SnarkBoojum** [comment]
I think I have something to add about H.Herbelin's comment : if you extract a software from whatever source files in a language Foo to sources in a Bar language -- then the sources in the Bar language are not the sources from your project.

So if you release a final product which links against a GPL software, any user can ask you for the sources, and you can't give only the Bar-language ones -- you will have to provide the Foo-language ones.
